src/org/mozilla/javascript/optimizer/BodyCodegen.java

```diff
@@ -3421,26 +3431,55 @@ private static boolean isArithmeticNode(Node node)
             || (type == Token.MUL);
     }
 
-    private void visitArithmetic(Node node, int opCode, Node child,
+    private void visitArithmetic(Node node, int type, Node child,
         Node parent)
     {
         int childNumberFlag = node.getIntProp(Node.ISNUMBER_PROP, -1);
         if (childNumberFlag != -1) {
             generateExpression(child, node);
             generateExpression(child.getNext(), node);
-            cfw.add(opCode);
+
+            switch (type) {
+            case Token.SUB:
+                cfw.add(ByteCode.DSUB);
+                break;
+            case Token.MUL:
+                cfw.add(ByteCode.DMUL);
+                break;
+            case Token.DIV:
+                cfw.add(ByteCode.DDIV);
+                break;
+            case Token.MOD:
+                cfw.add(ByteCode.DREM);
+                break;
+            default:
+                throw Kit.codeBug(Token.typeToName(type));
+            }
         }
         else {
             boolean childOfArithmetic = isArithmeticNode(parent);
             generateExpression(child, node);
             if (!isArithmeticNode(child))
-                addObjectToDouble();
+                addObjectToNumeric();
             generateExpression(child.getNext(), node);
             if (!isArithmeticNode(child.getNext()))
-                addObjectToDouble();
-            cfw.add(opCode);
-            if (!childOfArithmetic) {
-                addDoubleWrap();
+                addObjectToNumeric();
+
+            switch (type) {
+            case Token.SUB:
+                addScriptRuntimeInvoke("subtract", "(Ljava/lang/Number;Ljava/lang/Number;)Ljava/lang/Number;");
```

Comment:
I bet that this is where some of the performance regressions might be coming in -- before, in most cases (because I have rarely seen the "numeric optimization" above really work -- we convert both operands to a double and then do the operation in one bytecode. This replaces all of those with a call to a function that does a bunch of instanceof checks before invoking the instruction. 

I wonder if this would benchmark better if we first called a method that checked the operand types for big-integerness in the most efficient way we can, then either used the old code path or the new one. Eventually invokedynamic could be used to make the same thing happen.

Reply:
b3170b6

Reply:
Thanks. Sorry for the late reply.

I too have concerns about performance. However, I couldn't figure out what was slowed down by this change.

The bigint branch and master were each measured five times by `./gradlew testBenchmark`.
https://docs.google.com/spreadsheets/d/1sYM7EhGONY5-oQLFjuHwJou4DZEwJ7o9KCGB3-jKisc/edit?usp=sharing

There doesn't seem to be much difference except for `SunSpiderBenchmark.bitopsNsieveBits`. (`SunSpiderBenchmark.bitopsNsieveBits` is it getting faster?)

Reply:
Do you have any data or samples that show that using `invokedynamic` is faster?
How does `invokedynamic` work on Android?